### PR TITLE
Add @ClientFormParam to Reactive REST Client

### LINKED ISSUE #34477: Add @ClientFormParam to Reactive REST Client
### Description

This is analogous to [`@ClientQueryParam`](https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ClientQueryParam.java) but for [`@FormParam`](https://jakarta.ee/specifications/platform/10/apidocs/jakarta/ws/rs/formparam)s.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by Eng-Fouad**
Resolves #34477

### CODE REVIEW

**Review by Sgitario — CHANGES_REQUESTED** [review]
The changes look amazing to me. Many thanks!
The only missing part is to update the documentation similar to what is mentioned to `@ClientQueryParam`: https://quarkus.io/guides/rest-client-reactive#using-clientqueryparam

I'm also running the CI.

**Comment by Sgitario** [comment]
The test failures do seem related.

**Comment by Sgitario** [comment]
I'm having a look into the test failures

**Comment by Sgitario** [comment]
> I'm having a look into the test failures

The message was indeed misleading, though when running the test on debug mode, you add a breakpoint [here](https://github.com/quarkusio/quarkus/blob/8a0624822e522cd438781ee786355a1ce41d7f54/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java#L395), you will see the actual errors.

The problem is that the issue is found in the SubClient and that's why it could not identify the errors.

Anyway, the actual errors are:
- there are a couple of null pointers because formValues is null [here](https://github.com/quarkusio/quarkus/pull/34535/files#diff-3bf496e685c85459b6ad2b94694f39720676248db75e1bd2694cb9a21d9dc69bR541) and [here](https://github.com/quarkusio/quarkus/pull/34535/files#diff-3bf496e685c85459b6ad2b94694f39720676248db75e1bd2694cb9a21d9dc69bR435).
- the method definition in [here](https://github.com/quarkusio/quarkus/pull/34535/files#diff-3bf496e685c85459b6ad2b94694f39720676248db75e1bd2694cb9a21d9dc69bR131) is not correct, I think it should be:

```java
private static final MethodDescriptor MULTI_VALUED_MAP_ADD_ALL_METHOD = MethodDescriptor.ofMethod(MultivaluedMap.class,
            "addAll", void.class, Object.class, List.class);
```

After making these changes, still didn't work for me because the following exception at runtime:

```
java.lang.VerifyError: Bad local variable type
Exception Details:
  Location:
    io/quarkus/rest/client/reactive/form/ClientFormParamFromMethodTest$Client$$QuarkusRestClientInterface.setFromFormParam(Ljava/lang/String;)Ljava/lang/String; @116: aload
  Reason:
    Type top (current frame, locals[6]) is not assignable to reference type
  Current Frame:
    bci: @116
    flags: { }
    locals: { 'io/quarkus/rest/client/reactive/form/ClientFormParamFromMethodTest$Client$$QuarkusRestClientInterface', 'java/lang/String', 'java/lang/reflect/Type', '[Ljava/lang/annotation/Annotation;', 'jakarta/ws/rs/core/MultivaluedMap', top, top, top, top, top, top, top, 'jakarta/ws/rs/client/WebTarget' }
    stack: { }
  Bytecode:
```

But I hope you can continue investigating after my findings.

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/MicroProfileRestClientEnricher.java [hunk 1]
```diff
@@ -365,6 +422,140 @@ private void addQueryParam(MethodInfo declaringMethod, MethodCreator methodCreat
...
     }
 
+    private void addFormParam(MethodInfo declaringMethod, MethodCreator methodCreator,
+            ParamData paramData, BuildProducer<GeneratedClassBuildItem> generatedClasses,
+            IndexView index, AssignableResultHandle formValues) {
+
+        AnnotationInstance annotation = paramData.annotation;
+        ClassInfo declaringClass = paramData.definingClass;
+
+        String queryName = annotation.value("name").asString();
```
Comment (Sgitario) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/MicroProfileRestClientEnricher.java:432]: Please, rename it to `formName`

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/MicroProfileRestClientEnricher.java [hunk 2]
```diff
@@ -365,6 +422,140 @@ private void addQueryParam(MethodInfo declaringMethod, MethodCreator methodCreat
...
+        ResultHandle queryNameHandle = methodCreator.load(queryName);
```
Comment (Sgitario) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/MicroProfileRestClientEnricher.java:433]: Please, rename it to `formNameHandle`

**Review by Sgitario — CHANGES_REQUESTED** [review]
See my inline comments and last comment to fix the test failures.

**Comment by Eng-Fouad** [comment]
Merged with latest commit from main branch, and now I am getting new error:

```
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /mnt/c/Users/fouad/IdeaProjects/quarkus-fouad/extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:[531,24] method isImplementorOf in class org.jboss.resteasy.reactive.common.processor.JandexUtil cannot be applied to given types;
  required: org.jboss.jandex.IndexView,org.jboss.jandex.ClassInfo,org.jboss.jandex.DotName
  found:    org.jboss.jandex.CompositeIndex,org.jboss.jandex.ClassInfo,org.jboss.jandex.DotName,java.util.Set<org.jboss.jandex.DotName>
  reason: actual and formal argument lists differ in length
```

I think it is caused by #34570 cc @geoand

**Comment by geoand** [comment]
You need to rebased on main

**Comment by Eng-Fouad** [comment]
> You need to rebased on main and update the code

Already did that. Same error.

**Comment by Eng-Fouad** [comment]
> > I'm having a look into the test failures
> 
> The message was indeed misleading, though when running the test on debug mode, you add a breakpoint [here](https://github.com/quarkusio/quarkus/blob/8a0624822e522cd438781ee786355a1ce41d7f54/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java#L395), you will see the actual errors.
> 
> The problem is that the issue is found in the SubClient and that's why it could not identify the errors.
> 
> Anyway, the actual errors are:
> 
> * there are a couple of null pointers because formValues is null [here](https://github.com/quarkusio/quarkus/pull/34535/files#diff-3bf496e685c85459b6ad2b94694f39720676248db75e1bd2694cb9a21d9dc69bR541) and [here](https://github.com/quarkusio/quarkus/pull/34535/files#diff-3bf496e685c85459b6ad2b94694f39720676248db75e1bd2694cb9a21d9dc69bR435).
> * the method definition in [here](https://github.com/quarkusio/quarkus/pull/34535/files#diff-3bf496e685c85459b6ad2b94694f39720676248db75e1bd2694cb9a21d9dc69bR131) is not correct, I think it should be:
> 
> ```java
> private static final MethodDescriptor MULTI_VALUED_MAP_ADD_ALL_METHOD = MethodDescriptor.ofMethod(MultivaluedMap.class,
>             "addAll", void.class, Object.class, List.class);
> ```
> 
> After making these changes, still didn't work for me because the following exception at runtime:
> 
> ```
> java.lang.VerifyError: Bad local variable type
> Exception Details:
>   Location:
>     io/quarkus/rest/client/reactive/form/ClientFormParamFromMethodTest$Client$$QuarkusRestClientInterface.setFromFormParam(Ljava/lang/String;)Ljava/lang/String; @116: aload
>   Reason:
>     Type top (current frame, locals[6]) is not assignable to reference type
>   Current Frame:
>     bci: @116
>     flags: { }
>     locals: { 'io/quarkus/rest/client/reactive/form/ClientFormParamFromMethodTest$Client$$QuarkusRestClientInterface', 'java/lang/String', 'java/lang/reflect/Type', '[Ljava/lang/annotation/Annotation;', 'jakarta/ws/rs/core/MultivaluedMap', top, top, top, top, top, top, top, 'jakarta/ws/rs/client/WebTarget' }
>     stack: { }
>   Bytecode:
> ```
> 
> But I hope you can continue investigating after my findings.

Thanks for your help. I applied all fixes you mentioned. I will try to figure out how to fix `java.lang.VerifyError: Bad local variable type`.

**Comment by Eng-Fouad** [comment]
> You need to rebased on main and update the code

My bad. I had to recompile the project again.

**Comment by Eng-Fouad** [comment]
Ok, I fixed `java.lang.VerifyError: Bad local variable type`. However, the tests still fail:

```
[ERROR] Failures:
[ERROR]   ClientFormParamFromMethodTest.shouldUseValuesOnlyFromClass:39
expected: "1/"
 but was: "/"
[INFO]
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```

It seems that the form params are not passed in the request.

**Comment by Eng-Fouad** [comment]
Fixed. All tests should pass now.

**Review by Sgitario — CHANGES_REQUESTED** [review]
There are several concerns about the implementation:
- Generally speaking, we should be very careful when changing the signature of interfaces like JaxrsClientReactiveEnricher. In this case, we can avoid it (see the following points)
- `JaxrsClientReactiveEnricher.forWebTarget` and `JaxrsClientReactiveEnricher.forSubResourceWebTarget` is meant to enrich the webTarget and when handling the form params, we are not updating the web target. 
- You're using an array to hold a single instance of formParams (as if you were passing a reference parameter). 

To avoid these three points, I would create a new method in JaxrsClientReactiveEnricher like `handleFormParams` that returns the assignable result handle (to avoid the usage of arrays for doing this). 

Other comments:

- The method `MicroProfileRestClientEnricher.addFormParam` is too long and I think it can be refactored (eg reusing variables, etc).
- There are several conditions like `if (!multipart)`, I don't fully understand why you ended up with this condition, but form params are not incompatible with multipart. I guess your intention here is not to parse the annotations that are used in Multipart classes? If this scenario is not supported, we should throw an exception to let users that this is not a valid use case. If this was not your intention, I would need further clarification about this to better understand the condition.

**Comment by Eng-Fouad** [comment]
> There are several concerns about the implementation:
> 
> * Generally speaking, we should be very careful when changing the signature of interfaces like JaxrsClientReactiveEnricher. In this case, we can avoid it (see the following points)
> * `JaxrsClientReactiveEnricher.forWebTarget` and `JaxrsClientReactiveEnricher.forSubResourceWebTarget` is meant to enrich the webTarget and when handling the form params, we are not updating the web target.
> * You're using an array to hold a single instance of formParams (as if you were passing a reference parameter).
> 
> To avoid these three points, I would create a new method in JaxrsClientReactiveEnricher like `handleFormParams` that returns the assignable result handle (to avoid the usage of arrays for doing this).
> 

I agree with you, using separate method `handleFormParams` is much better. I will do that.

> Other comments:
> 
> * The method `MicroProfileRestClientEnricher.addFormParam` is too long and I think it can be refactored (eg reusing variables, etc).

Do you mean like merging `addFormParam` with `addQueryParam` (e.g. new method `addParam`) and then pass method argument that specifies param type?

> * There are several conditions like `if (!multipart)`, I don't fully understand why you ended up with this condition, but form params are not incompatible with multipart. I guess your intention here is not to parse the annotations that are used in Multipart classes? If this scenario is not supported, we should throw an exception to let users that this is not a valid use case. If this was not your intention, I would need further clarification about this to better understand the condition.

In case of multipart, `forParams` is `QuarkusMultipartForm` not `MultivaluedHashMap`. With `MultivaluedHashMap`, we can simply add it as follows:

```
forParams.addAll(annotation.name(), Arrays.asList(annotation.value()));
```

However, with `QuarkusMultipartForm`, should I use `attribute(String name, String value, String filename)`? How to construct `filename`?

**Comment by Eng-Fouad** [comment]
@geoand @Sgitario Any updates on this PR? Thanks.

**Comment by geoand** [comment]
@Sgitario is on PTO, so let's wait for him to have the final say

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/MicroProfileRestClientEnricher.java [hunk 3]
```diff
@@ -184,70 +201,178 @@ public void forSubResourceWebTarget(MethodCreator methodCreator, IndexView index
...
             AssignableResultHandle webTarget, BuildProducer<GeneratedClassBuildItem> generatedClasses) {
 
-        Map<String, QueryData> queryParamsByName = new HashMap<>();
-        collectClientQueryParamData(rootInterfaceClass, rootMethod, queryParamsByName);
-        collectClientQueryParamData(subInterfaceClass, subMethod, queryParamsByName);
+        Map<String, ParamData> queryParamsByName = new HashMap<>();
+        collectClientParamData(rootInterfaceClass, rootMethod, queryParamsByName,
+                CLIENT_QUERY_PARAM, CLIENT_QUERY_PARAMS, "ClientQueryParam");
+        collectClientParamData(subInterfaceClass, subMethod, queryParamsByName,
+                CLIENT_QUERY_PARAM, CLIENT_QUERY_PARAMS, "ClientQueryParam");
```
Comment (Sgitario) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/MicroProfileRestClientEnricher.java:208]: Please, move "ClientQueryParam" into a constant and/or provide a custom collectClientParamData method only for query params.

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/MicroProfileRestClientEnricher.java [hunk 4]
```diff
@@ -184,70 +201,178 @@ public void forSubResourceWebTarget(MethodCreator methodCreator, IndexView index
...
-            headerFillersByName.put(classLevelHeader.value("name").asString(),
-                    new QueryData(classLevelHeader, interfaceClass));
+    @Override
+    public AssignableResultHandle handleFormParams(MethodCreator methodCreator, IndexView index, ClassInfo interfaceClass,
+            MethodInfo method, BuildProducer<GeneratedClassBuildItem> generatedClasses, AssignableResultHandle formParams,
+            boolean multipart) {
+
+        Map<String, ParamData> formParamsByName = new HashMap<>();
+        collectClientParamData(interfaceClass, method, formParamsByName,
+                CLIENT_FORM_PARAM, CLIENT_FORM_PARAMS, "ClientFormParam");
```
Comment (Sgitario) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/MicroProfileRestClientEnricher.java:222]: Please, move "ClientFormParam" into a constant and/or provide a custom collectClientParamData method only for form params.

**Review by Sgitario — COMMENTED** [review]
Back from my PTO. Only two comments to address. 
The rest of the changes look good to me. Thanks!

**Comment by Eng-Fouad** [comment]
> Back from my PTO. Only two comments to address. The rest of the changes look good to me. Thanks!

Welcome back. I updated the code according to your comments. Thanks.

**Review by Sgitario — APPROVED** [review]

**Pull request merged**